EIPS/eip-7937.md

```diff
@@ -0,0 +1,145 @@
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Prefix opcode behavior
+
+This EIP uses the prefix opcode `C0`, and it only occupies this single EVM opcode space. Upon the interpreter encountering opcode `C0`, it MUST continue to seek the next byte in code. It then executes things in "64-bit mode", based on the second byte, described below. If the execution is successful, then the interpreter MUST increase `PC` by 2 (instead of 1).
+
+If the second byte is not a valid 64-bit mode operation, then the interpreter MUST OOG.
+
+### General 64-bit mode behavior
+
+In 64-bit mode, all operations only work on the lowest 64-bit of each stack value. The first 192-bit is discarded. When a result value is pushed back onto the stack, then it MUST ensures that observable effects will see that the first 192-bit is zero. Note that here it's not necessary for an interpreter to reset the first 192-bit to zero every time -- if the next opcode is still in 64-bit mode, then the first 192-bit is still unobservable. We discuss the full details in the "rationale" section. The interpreter only needs to reproduce the full 256-bit value upon entering non-64-bit mode. If the full computational heavy part can be written in pure 64-bit mode, then this can result in noticable performance gain.
+
+### Gas cost constants
+
+We define the following gas cost constants:
+
+* `G_BASE64`: 1
+* `G_VERYLOW64`: 2
+* `G_LOW64`: 3
+* `G_MID64`: 4
+
+### Arithmetic opcodes
+
+The 64-bit mode arithmetic opcodes are defined the same as non-64-bit mode, except that it only operates on the last 64-bits. In the below definition, `a`, `b`, `N` is `a mod 2^64`, `b mod 2^64` and `N mod 2^64`.
+
+* ADD (`C001`) and SUB (`C003`): `a op b mod 2^64`, gas cost `G_VERYLOW64`.
+* MUL (`C002`), DIV (`C004`), SDIV (`C005`), MOD (`C006`), SMOD (`C007`), SIGNEXTEND (`C00B`): `a op b mod 2^64`, gas cost `G_LOW64`.
+* ADDMOD (`C008`), MULMOD (`C009`): `a op b % N mod 2^64`, gas cost `G_MID64`.
+* EXP (`C00A`): `a EXP b mod 2^64`, gas cost is the same as non-64-bit mode.
+
+### Comparison and bitwise opcodes
+
+The 64-bit mode comparison and bitwise opcodes are defined the same as non-64-bit mode, except that they only operates on the last 64-bits.
+
+* LT (`C010`), GT (`C011`), SLT (`C012`), SGT (`C013`), EQ (`C014`), AND (`C016`), OR (`C017`), XOR (`C018`): `a op b mod 2^64`, gas cost `G_VERYLOW64`
+* ISZERO (`C015`), NOT (`C019`): `op a mod 2^64`, gas cost `G_VERYLOW64`
+
+Note that:
+
+* 64-bit EQ (`C014`) may return true for two different integers because it'll only compare the last 64-bits.
```

Comment:
Ok, this is confusing, regarding my point of SHR just now. How does SHR work if there are topmost bits? This seems to imply I can smuggle top bits into my 64 bit mode? Is this intended?